Add check for IPv6 before generating bpf headers

christarazi opened the pull request:
When IPv6 is disabled (passing --enable-ipv6=false), this commit prevents writing IPv6-related data to bpf header files.

The files that are affected from this change are the following:

```
<cilium state dir>/globals/node_config.h
<cilium state dir>/<endpoint ID>/lxc_config.h
```

In theory we would need to make a similar change for when IPv4 is disabled, but that is not likely in practice. It is something we can explore in the future.

Fixes #8032

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
Add check for IPv6 before generating bpf headers

When IPv6 is disabled (passing --enable-ipv6=false), this commit
prevents writing IPv6-related data to bpf header files. Note that this
does not include the template bpf header file as we want to ensure that
the dummy IPv6 address is present, hence being a template.

The files that are affected from this change are the following:

<cilium state dir>/globals/node_config.h
<cilium state dir>/<endpoint ID>/lxc_config.h

Fixes #8032

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/datapath/linux/config/config.go:
```diff
@@ -109,8 +113,10 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
-	extraMacrosMap["HOST_IP"] = hostIP.String()
-	fw.WriteString(defineIPv6("HOST_IP", hostIP))
+	if option.Config.EnableIPv6 {
+		extraMacrosMap["HOST_IP"] = hostIP.String()
```
I don't think this line is suppose to be conditional by the `option.Config.EnableIPv6`

joestringer replied:
It's only used in `ENABLE_IPV6` paths:

```
$ git grep HOST_IP bpf/
bpf/lib/encap.h:                BPF_V6(host_ip, HOST_IP);
bpf/lib/encap.h:                BPF_V6(host_ip, HOST_IP);
bpf/node_config.h:DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x0, 0x2, 0xf, 0xff, 0xff);
```

joestringer reviewed (changes requested):
What's here looks good, but looking at CI I think that the template substitution pieces are relying on the LXC_IP being defined in the binary even if IPv6 is disabled. 

Following the ginkgo link at the bottom of the PR:
* https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/18032/
* Click on one of the test results
* https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/18032/testReport/junit/Suite-k8s-1/11/K8sDatapathConfig_DirectRouting_Check_direct_connectivity_with_per_endpoint_routes/
* Attachments includes sysdump of the cluster when the failure occurred
* Download, open zip
* Digging in, there are `pod-kube-system-cilium-....` logs
* Grepping in there for warnings/errors finds this:
```
2020-03-19T04:43:23.404234718Z level=error msg="endpoint regeneration failed" containerID=7b32e6fdfc datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=764 error="failed to write ELF file: 764_next/bpf_lxc.o: no such symbol \"LXC_IP_1\" in ELF" identity=36003 ipv4=10.10.0.135 ipv6="f00d::a0b:0:0:e3dc" k8sPodName=external-ips-test/app1-798d4f944d-qpzpf subsys=endpoint
```

christarazi committed:
Update unit tests for WriteEndpointConfig

This commit adds coverage for EnableIPv6 when generating
{template,endpoint} bpf header files and programs.

These tests ensure that regardless of the value of EnableIPv6, the
template bpf header files and resulting object file have a dummy IPv6
address as a #define "DEFINE_IPV6" and present as a symbol
LXC_IP_{1,2,3,4}, respectively.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer commented on pkg/datapath/linux/config/config_test.go:
```diff
@@ -108,6 +110,88 @@ func (s *ConfigSuite) TestWriteEndpointConfig(c *C) {
+			template:    testutils.NewTestEndpoint(),
+			endpoint:    testutils.NewTestEndpoint(),
+			preTestRun: func(t *testutils.TestEndpoint, e *testutils.TestEndpoint) {
+				option.Config.EnableIPv6 = false
```
@cilium/ci folks, is this best practice or do you have an alternative suggestion here?

christarazi replied:
I can see how that could pollute the other tests. One way could be to set the option back to its original value at the end of the function.

raybejjani replied:
The leakage would be my primary concern. I don't know if we have too much by way of best-practice in the unit tests. I suspect that they are scoped to each package so you shouldn't be re-initialising stuff out-of-order with this setup. What don't know is what happens when someone realises that we can run unit tests in parallel.

Is it possible to simply change how the components we test are initialised so that they take the ipv6 parameter instead of wiring in the global?

christarazi replied:
Hmm, it is possible, but my quick judgements say that it would require a major refactor. 

Edit: You can specify which tests you'd like to parallelize via [`t.Parallel()`](https://golang.org/pkg/testing/#T.Parallel). Golang will only parallelize tests that are explicitly marked. One potential transitional approach could be to mark tests that are deemed safe to parallelize, while we work on a refactor (if the refactor is valuable to take on).

joestringer replied:
Sounds like restoring the option at the end of the test is a reasonable middle-ground.

We can keep thinking about the best way to represent the options throughout packages going forward with this in mind.

christarazi replied:
Done

joestringer reviewed (approved):
Looks good to me, would like feedback from CI folks on one point.
